### ISSUE TITLE
Question: indexer method that returns "ref T" doesn't get inlined.

### ISSUE THREAD

**Issue by hypeartist** [dotnet/coreclr#15457]
Checked under: **Release Build, .Net Framework v4.7.1, VS 2017 v15.5.1, System.Memory v4.4.0-preview1-25305-02, System.Runtime.CompilerServices.Unsafe v4.4.0**

Some data container class with indexer
```csharp
public sealed unsafe class CellArray1
{
	private Cell* _data;

	public ref Cell this[int pos]
	{
		[MethodImpl(MethodImplOptions.AggressiveInlining)]
		get { return ref *(_data + pos); }
	}
}
```
Some silly  loop
```csharp
var m1 = new CellArray1();
m1.Resize(4096);

var sum1 = 0;

for (var i = 0; i < 4096; i++)
{
	m1[i].X = i;
	m1[i].Y = -i;
	sum1 += m1[i].X;
}

m1.Dispose();
```
IL:
```IL
    // start of loop, entry point: IL_009c

      // [53 4 - 53 5]
      IL_0063: nop          

      // [58 5 - 58 17]
      IL_0064: ldloc.2      // m1
      IL_0065: ldloc.s      i
      IL_0067: callvirt     instance valuetype [TestLib]TestLib.Cell& TestLib.CellArray1::get_Item(int32)
      IL_006c: ldloc.s      i
      IL_006e: stfld        int32 [TestLib]TestLib.Cell::X

      // [59 5 - 59 18]
      IL_0073: ldloc.2      // m1
      IL_0074: ldloc.s      i
      IL_0076: callvirt     instance valuetype [TestLib]TestLib.Cell& TestLib.CellArray1::get_Item(int32)
      IL_007b: ldloc.s      i
      IL_007d: neg          
      IL_007e: stfld        int32 [TestLib]TestLib.Cell::Y

      // [60 5 - 60 21]
      IL_0083: ldloc.s      sum1
      IL_0085: ldloc.2      // m1
      IL_0086: ldloc.s      i
      IL_0088: callvirt     instance valuetype [TestLib]TestLib.Cell& TestLib.CellArray1::get_Item(int32)
      IL_008d: ldfld        int32 [TestLib]TestLib.Cell::X
      IL_0092: add          
      IL_0093: stloc.s      sum1

      // [69 4 - 69 5]
      IL_0095: nop          

      // [52 30 - 52 33]
      IL_0096: ldloc.s      i
      IL_0098: ldc.i4.1     
      IL_0099: add          
      IL_009a: stloc.s      i

      // [52 20 - 52 28]
      IL_009c: ldloc.s      i
      IL_009e: ldc.i4       4096 // 0x00001000
      IL_00a3: clt          
      IL_00a5: stloc.s      V_10

      IL_00a7: ldloc.s      V_10
      IL_00a9: brtrue.s     IL_0063
    // end of loop
```
Code generated by JIT :
```asm
000007FE84770517  xor         r14d,r14d  
000007FE8477051A  mov         rcx,rsi  
000007FE8477051D  mov         edx,r14d  
000007FE84770520  call        000007FE84770150 // <---- ?
000007FE84770525  mov         dword ptr [rax],r14d  
000007FE84770528  mov         rcx,rsi  
000007FE8477052B  mov         edx,r14d  
000007FE8477052E  call        000007FE84770150  // <---- ?
000007FE84770533  mov         ecx,r14d  
000007FE84770536  neg         ecx  
000007FE84770538  mov         dword ptr [rax+4],ecx  
000007FE8477053B  mov         rcx,rsi  
000007FE8477053E  mov         edx,r14d  
000007FE84770541  call        000007FE84770150 // <---- ?  
000007FE84770546  add         ebp,dword ptr [rax]  

000007FE84770548  inc         r14d  
000007FE8477054B  cmp         r14d,1000h  
000007FE84770552  jl          000007FE8477051A
```
And here is the example without using **ref T**

Some data container class with indexer
```csharp
public sealed unsafe class CellArray2
{
	private Cell* _data;

	public Cell* this[int pos]
	{
		[MethodImpl(MethodImplOptions.AggressiveInlining)]
		get { return _data + pos; }
	}
}
```
Some silly  loop
```csharp
var m1 = new CellArray2();
m2.Resize(4096);

var sum2 = 0;

for (var i = 0; i < 4096; i++)
{
	m2[i]->X = i;
	m2[i]->Y = -i;
	sum2 += m2[i]->X;
}

m1.Dispose();
```
Code generated by JIT :
```asm
000007FE84780515  xor         ecx,ecx  
000007FE84780517  mov         rax,qword ptr [rdi+8]  
000007FE8478051B  movsxd      rdx,ecx  
000007FE8478051E  lea         rdx,[rdx+rdx*2]  
000007FE84780522  mov         dword ptr [rax+rdx*8],ecx  
000007FE84780525  mov         rax,qword ptr [rdi+8]  
000007FE84780529  mov         r8d,ecx  
000007FE8478052C  neg         r8d  
000007FE8478052F  mov         dword ptr [rax+rdx*8+4],r8d  
000007FE84780534  mov         rax,qword ptr [rdi+8]  
000007FE84780538  add         ebp,dword ptr [rax+rdx*8]  

000007FE8478053B  inc         ecx  
000007FE8478053D  cmp         ecx,1000h  
000007FE84780543  jl          000007FE84780517 
```

**Comment by hypeartist** [comment]
Just tried using **System.Runtime.CompilerServices.Unsafe**

Some data container class with indexer:

```csharp
public sealed unsafe class UnsafeArray<T> : where T : struct
{
	private static readonly int _itemSize = Unsafe.SizeOf<T>();
	private byte* _data;

	public ref T this[int pos] => ref Unsafe.AsRef<T>(_data + pos * _itemSize);
}
```
Some silly loop:
```csharp
var m4 = new UnsafeArray<Cell>();
m4.Resize(4096);

var sum4 = 0;

for (var i = 0; i < 4096; i++)
{
	m4[i].X = i;
	m4[i].Y = -i;
	sum4 += m4[i].X;
}
```
Code generated by JIT :
```asm
000007FE85351094  xor         ecx,ecx  
000007FE85351096  lea         eax,[rcx+rcx*2]  
000007FE85351099  shl         eax,3  
000007FE8535109C  movsxd      rax,eax  
000007FE8535109F  mov         rdx,rax  
000007FE853510A2  add         rdx,qword ptr [rbp+8]  
000007FE853510A6  mov         dword ptr [rdx],ecx
000007FE853510A8  mov         rdx,rax  
000007FE853510AB  add         rdx,qword ptr [rbp+8]  
000007FE853510AF  mov         r8d,ecx  
000007FE853510B2  neg         r8d  
000007FE853510B5  mov         dword ptr [rdx+4],r8d 
000007FE853510B9  add         rax,qword ptr [rbp+8]  
000007FE853510BD  add         r14d,dword ptr [rax]

000007FE853510C0  inc         ecx  
000007FE853510C2  cmp         ecx,1000h  
000007FE853510C8  jl          000007FE85351096
```

**Comment by jkotas** [comment]
This was code optimization issue in the JIT. This issue is fixed in .NET Core, and the fix should show up in one of the future .NET Framework updates as well.

As you have noticed, System.Runtime.CompilerServices.Unsafe has a workaround for it: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il#L277

**Comment by AndyAyersMS** [comment]
For future reference: the inliner tracks the reasons for each failed inline back and these can be surfaced in a variety of ways
* Enable inlining events in ETW and view the resulting ETL. For instance via a tool like [perfview](https://github.com/Microsoft/perfview).
* The [inlining analyzer](https://github.com/szehetner/InliningAnalyzer) is a new add-in tool that can show inlining failures and successes right in your source code within the Visual Studio IDE. I couldn't get it to work for properties, but here it is on your example if I convert to a regular method:

![image](https://user-images.githubusercontent.com/10121823/33895493-7c0c9040-df15-11e7-8433-f0fbb62e5263.png)


* CoreCLR only: If you can build a matching CHK version of clrjit.dll and copy it in place, you can tell the jit to display inlines by setting `COMPlus_JitPrintInlinedMethods=1`.

As Jan mentions the jit inlining limitation is fixed in Core 2.0 and the fix will be in Desktop 4.7.2 (looks like it maybe just missed being in 4.7.1). See #11218.

**Comment by hypeartist** [comment]
Thanks, @jkotas, @AndyAyersMS, for clarification!
